docs/proposals/api-grouping.md

```diff
@@ -0,0 +1,375 @@
+  would not accept.  We as a core project have to use the extension mechanism to keep
+  ourselves honest, though.
+
+- The same reasons programs have packages: to enforce boundaries, and prevent coupling (taken from #635).
+- To allow api versioning at different rates (taken from #635)
+  - particularly for hosted implementations
+- To allow replacement of project-provided kinds with proprietary types
+  - PaaSes on top of Kubernetes might to do this.
+- So that users can focus on learning one group at a time.
+- So that security/policy admins can audit/enable an API group at a time, and not need to
+  review all kinds at once.
+- So that different groups within the kubernetes project can have a greater degree of autonomy
+  - Conways law 
+  - have trusted api reviewers per group.
+
+## Ways to Group things
+
+Listed below are some ways that you might want to group things.
+It is not a proposal to do all those things; in fact some are in conflict.
+
+### Proposal from #635
+
+Issue #635 proposed this grouping:
+
+```
+apis/podScheduling/v1
+	Pods
+	PodTemplates
+	Bindings
+
+apis/services/v1
+	Services
+
+apis/rcs/v1
+	ReplicationController
+
+apis/nodes/v1
+	Nodes
+```
+
+### Admin vs User
+
+Issue #3806 proposed splitting "Cluster" and
+"User" resources.  Applied to the current state,
+that would look something like this:
+
+```
+/api/v1
+	persistentVolumeClaim
+	pod
+	podTemplate
+	replicationController
+	service
+	serviceAccount
+	endpoints
+	event
+	secret
+	petSet
+	workflow
+	scheduledJob
+	componentStatus   # Maybe goes in apis/cluster?
+	ingress           # Maybe goes in apis/cluster?
+	podSecurityPolicyBinding
+
+/apis/metrics/v1alpha1
+	Pod
+
+/apis/cluster/v1
+	node
+	namespace
+	binding
+	persistentVolume
+	limitRange
+	resourceQuota
+	daemonSet
+	clusterAutoscaler
+	metadataPolicy
+	podSecurityPolicy
+
+/apis/clustermetrics/v1alpha1
+	Pod
+	Node
+
+/apis/extensions/v1beta1
+	scale
+	horizontalPodAutoscaler
+	thirdPartyResource
+	thirdPartyResourceData
+	deployment
+	job
+	replicaSet
+	scaleStatus
+	scale
+```
+
+Evaluation:
+
+- Everything still in the main API (except componentStatus and maybe podSecurityPolicy)
+  is stuff that makes sense to namespace, since they are things you replicate if
+  you make a "copy" of a service or system.
+- Everything in the `apis/cluster/v1` group does not need a namespace,
+  except clusterAutoscaler, limitRange, and resourceQuota
+- Everything in `apis/cluster/v1` is ostensibly stuff that a tenant does not need
+  to view or modify to use the system.  And it is stuff that the a hosted
+  kubernetes provider certainly wants tight control over, and may want
+  to hide so that it can be implemented differently.
+- non-admins can focus their learning on the User group.
+
+### Vertical Grouping
+
+Group things that work with the same type of workloads.
+
+Sketch:
+```
+/apis/batch/v1
+	scheduledJob
+	workflow
+	job
+
+/apis/longrunning/v1  # need better name
```

Comment:
I just picked "apps" for the first cut of petset. I know what "batch" means because I'm used to hearing it, just like replicated, clustered etc, not sure that applies for "continuous"; I think it might be confusing without included context. We already have "autoscaling".